### ISSUE TITLE
Does it support the model wrapped by DistributedDataParallel？

### ISSUE THREAD

**Issue by guozhiyao** [learnables/learn2learn#197]

**Comment by seba-1511** [comment]
Yes it does, but that requires a bit of care. We'll push something out soon to make this much easier.

**Comment by guozhiyao** [comment]
> Yes it does, but that requires a bit of care. We'll push something out soon to make this much easier.

@seba-1511 Could you share a demo about how to wrap the model with DDP? Should I wrap the model or the maml with DDP?

**Comment by brando90** [comment]
@seba-1511 I am curious, does learn2learn work with DDP?

I was trying to speed up my training (but I use torch meta) and I noticed that DDP might not work with their code. I've spent some time researching how to make it work and I believe I could make it work with [pytorch's RPC](https://pytorch.org/tutorials/intermediate/rpc_tutorial.html) but not being an expert in distributed training I am wondering if I am even doing the right approach or if I am overcomplicating things. (numbering for each of refernce) 
1. Do you think using RPC is overkill? 
2. Does DDP work with learn2learn?

**Comment by seba-1511** [comment]
@brando90 If you're doing supervised meta-learning and want to parallelize over task iterations, you can do that with our lightning wrappers. Have a look at our doc ([here](http://learn2learn.net/docs/learn2learn.algorithms/)) and at the lightning documentation regarding DDP. (Tested with lightning==1.0.2)

To answer your questions:

1. Probably overkill -- unless you have a huge network but even then I'd favor DataParallel.
2. Yes it works -- either via lightning or you can also write your own custom DDP training script.

**Comment by brando90** [comment]
@seba-1511 Hi Seba! Thanks for your quick reply.

I will definitively take a look at your doc and the lighting wrapper with regard to DDP. 

In response to your reply I will number again. 

1. When you say you'd favor DataParallel - I assume you mean DDP and not DataParallel. Right? I am interested in multiprocessor (with or without gpu) but I am not splitting my model across machines nor GPUs.
2. I was wondering, I also have custom meta-L data sets (and torchmeta dataloaders). I was wondering, does your framework support using torchmeta? Or perhaps I can use my raw data somehow around your library? Do you have a pointer?
3. I am also currently using my own implementation of MAML using higher. Will learn2learn's DDP/lighting packages support using higher? i.e. will the backward pass be computed correctly even with higher in my code?

Thanks again for your time. It's highly appreciated.

**Comment by seba-1511** [comment]
1. Yes, I mean DataParallel but given your answer I don't think it'll help.
2. learn2learn provides datasets and algorithms, both independent of the other. For an example of MAML + torchmeta, [check this](https://github.com/learnables/learn2learn/blob/master/examples/maml_sine.py).
3. Our implementation of MAML is orthogonal to higher (same result, different paths). If you're using vanilla MAML, this [high-level interface](http://learn2learn.net/docs/learn2learn.algorithms/#maml) is well-tested and fairly flexible (we use it for SL & RL).
4. Yes, PyTorch Lightning.

I saw you've been looking for a distributed MAML implementation for a while. Given how tricky MAML can be, I would try to average the meta-gradients with `torch.distributed` directly instead of using DDP. The distributed optimizer in cherry might help for that: https://github.com/learnables/cherry/blob/master/cherry/optim.py

**Comment by brando90** [comment]
> 1. Yes, I mean DataParallel but given your answer I don't think it'll help.
> 2. learn2learn provides datasets and algorithms, both independent of the other. For an example of MAML + torchmeta, [check this](https://github.com/learnables/learn2learn/blob/master/examples/maml_sine.py).
> 3. Our implementation of MAML is orthogonal to higher (same result, different paths). If you're using vanilla MAML, this [high-level interface](http://learn2learn.net/docs/learn2learn.algorithms/#maml) is well-tested and fairly flexible (we use it for SL & RL).
> 4. Yes, PyTorch Lightning.
> 
> I saw you've been looking for a distributed MAML implementation for a while. Given how tricky MAML can be, I would try to average the meta-gradients with `torch.distributed` directly instead of using DDP. The distributed optimizer in cherry might help for that: https://github.com/learnables/cherry/blob/master/cherry/optim.py

@seba-1511 thank you so much for your help Seba, I appreciate it a lot.

I think we are getting very close for me to get the solution I've been looking for. 
 
What I think I am confused is why you are suggesting `torch.distributed` instead of DDP. I had the impression that pytorch-lighting gave parallelization out of the box, so I expected that your suggestion to use pytorch lighting would have been enough (no need for DDP or torch.distributed or cherry). Given the new suggestion I am slightly confused which one to try. Is the one you suggest the most is to use the cherry distributed optimizer or learn2learn's Lightning interface?

Thanks again for your time.

ps: btw, your tutorials I found on the pytorch forum were great! thanks for sharing that stuff with us :)

**Comment by brando90** [comment]
> 2\. learn2learn provides datasets and algorithms, both independent of the other. For an example of MAML + torchmeta, [check this](https://github.com/learnables/learn2learn/blob/master/examples/maml_sine.py).

btw, on the side. Thanks for this reply but now I realize that torchmeta won't work for me given that it doesn't work with [DDP](https://github.com/tristandeleu/pytorch-meta/issues/100). Unless I realized, that cherry's distributed optimizer doesn't interact at all with torchmeta...

**Comment by seba-1511** [comment]
Thanks for the nice words!

Correct, there are 2 ways to go about distributed MAML: either use DDP + LightningMAML, or use torch.distributed + cherry. Choose the one you're more familiar with :) As for torchmeta, I'm not sure if it's compatible with cherry.

**Comment by brando90** [comment]
> Thanks for the nice words!
> 
> Correct, there are 2 ways to go about distributed MAML: either use DDP + LightningMAML, or use torch.distributed + cherry. Choose the one you're more familiar with :) As for torchmeta, I'm not sure if it's compatible with cherry.

Thanks so much for your help! I really do appreciate it. I've been a bit frustrated for the past week and having to digest that all the code I had with higher and torchmeta is essentially useless for me now. Oh well! A walk and some nice food might make my grieving pass.

I am actually not familiar with pytorch lightning - plus I want to have access to the adapted models and not adapted models at the granularity of task - so, let me ask you if this would: Learn2LearnMAML ([no lightining like in this example](https://github.com/learnables/learn2learn/blob/master/examples/vision/maml_miniimagenet.py)) + DDP would work?

I wish I could do the torch.distributed + cherry way (sounds like I'd have more control over my code), but unfortunately, I do not know how to split the meta_batch into chunks and feed it into each maml instance on each processor, but it sounds that DDP would do that for me so I will try that first as it seems more achievable given my current skills on pytorch distributed training.

(funny I actually feel more comfortable with pytorch RPC autograd than cherry + torch.distributed...since I thought that would work with higher. If I get it to work I am happy to contribute it to your library! Thanks again! My words are not only nice but the truth ;) )

**Comment by SungFeng-Huang** [comment]
> I am actually not familiar with pytorch lightning - plus I want to have access to the adapted models and not adapted models at the granularity of task - so, let me ask you if this would: Learn2LearnMAML ([no lightining like in this example](https://github.com/learnables/learn2learn/blob/master/examples/vision/maml_miniimagenet.py)) + DDP would work?

Hi, I've successfully implemented Learn2Learn's LightningMAML + PyTorch Lightning DDP. Please check out my example code here: https://gist.github.com/SungFeng-Huang/dec22eef5650f5a74d24a732ffd0080f, it should work with adding the argument "--meta_task_ddp".

**Comment by arijitthegame** [comment]
I am sorry to open this issue up. But I was wondering if anyone wrote DDP wrappers to wrap l2l datasets. I am trying to use DDP with the ANIL algorithm as in the example folder. Thanks again.

**Comment by brando90** [comment]
related: https://github.com/learnables/learn2learn/issues/263#issuecomment-963372085

**Comment by brando90** [comment]
USE THIS for distributed MAML: https://github.com/learnables/learn2learn/issues/263#issuecomment-963372085
